ext/socket/raddrinfo.c

```diff
@@ -477,6 +483,26 @@ do_pthread_create(pthread_t *th, void *(*start_routine) (void *), void *arg)
     return ret;
 }
 
+static void *
+async_getaddrinfo(void *ptr)
+{
+    struct getaddrinfo_arg *arg = (struct getaddrinfo_arg *)ptr;
+    pthread_t th;
+    if (do_pthread_create(&th, do_getaddrinfo, arg) != 0) {
+        return (void *)EAI_AGAIN;
+    }
+    pthread_detach(th);
+
+    rb_thread_call_without_gvl2(wait_getaddrinfo, arg, cancel_getaddrinfo, arg);
+    return NULL;
+}
+
+static void *
+fork_safe_async_getaddrinfo(void *ptr)
+{
+    return rb_thread_prevent_fork(async_getaddrinfo, ptr);
```

Comment:
@akr pointed out that it is wrong to call `rb_thread_prevent_fork` here.

If Ruby's getaddrinfo is interrupted, such as by Ctrl+C, the pthread in question calling getaddrinfo(3) may remain even after `async_getaddrinfo` returns. So this does not satisfy your purpose (to avoid `fork` is called during getaddrinfo(3) execution).
Instead, you should guard the call of getaddrinfo(3) (and freeaddrinfo(3) too?), in do_getaddrinfo.

Note that `rb_thread_prevent_fork` would be called with no GVL in this case. Because `rb_ensure` requires GVL, `rb_thread_prevent_fork` must not use it. I think `rb_thread_prevent_fork` doesn't have to guard longjmp.

Reply:
Thank you @mame. I finally figured my linking error, and made changes based on your review.

I think it's much simpler now, how does it look?

Reply:
@casperisfine Thanks! It looks good to me. @akr Could you review it too?

Reply:
@akr said it looks okay

Reply:
That's great. Does that means you are good with merging it, or do you wish to add it at a dev meeting first? https://bugs.ruby-lang.org/issues/20590

Reply:
Let's merge it!

Reply:
🎉 I rebased it and enabled auto-merge. Thanks again!